General record member referencing (#3516)
modelica/ModelicaSpecification: chapters/operatorsandexpressions.tex

\end{center}
\end{table}

The postfix array index and postfix access operators can both be part of a \lstinline[language=grammar]!component-reference! (one of the alternative productions for \lstinline[language=grammar]!primary! in the grammar) and be applied to general expressions when the left operand is parenthesized.

> Directly using both postfix access operator and postfix array index in a \lstinline[language=grammar]!component-reference! has a special intuitive meaning.
> See \cref{indexing} and \cref{member-access-operator}.

Reviewer: I think we can leave the intuitiveness for the reader to decide:
```suggestion
There are semantic differences between the \lstinline[language=grammar]!component-reference! forms and the parenthesized forms of the operators, see \cref{indexing} and \cref{member-access-operator}.
```

Author: ```suggestion
Directly using both postfix access operator and postfix array index in a \lstinline[language=grammar]!component-reference! has a special meaning.
See \cref{indexing} and \cref{member-access-operator}.
```
I would then propose to just remove "intuitive" as above.

The key point is that only the case where you combine the two operators has a special meaning for component-references, and that is not syntactically legal for the parenthesized forms.

Reviewer: Sounds good.

The postfix array index and postfix access operators are not merely expression operators in the normal sense that \lstinline!a.b[1]! can be treated as \lstinline!a.(b[1])!.
Instead, these operators need to be considered jointly to identify an entire \lstinline[language=grammar]!component-reference! (one of the alternative productions for \lstinline[language=grammar]!primary! in the grammar) which is the smallest unit that can be seen as an expression in itself.

> Postfix access can only immediately be applied to a \lstinline[language=grammar]!component-reference!.

Reviewer: This contradicts the previous explanation that postfix access is part of the component reference.

Reviewer: I find the previous two sentences (_The postfix array index…_ and _Instead these operators…) problematic, and suggest that we replace them with something like this:
```
The postfix array index and postfix access operators can both be part of a \lstinline[language=grammar]!component-reference! (one of the alternative productions for \lstinline[language=grammar]!primary! in the grammar) and be applied to general expressions when the left operand is parenthesized.
```

Author: I added that, and then realized that the preceding text needed an update as well - so I rewrote that.
I hope it makes more sense now.

Reviewer: Even though I'm not exactly sure what changes to the _preceding test_ that correspond to this comment, I can't immediately spot any problem with the current formulations related to this.  Resolving.

> Postfix access can additionally be applied if there are parentheses around the left operand, see \cref{member-access-operator}.
> Postfix array index can additionally be applied if there are parentheses around the left operand, see \cref{indexing}.

Reviewer: Just need to keep cross references:
```suggestion
See \cref{indexing} and \cref{member-access-operator}.
```

(a.x)[2]    // OK: Component reference of type Real[2] - same as a[2].x[:]
(a.x)[2, :] // OK: Component reference of type Real[2] - same as a[2].x[:]
a[3]        // OK: Component reference of type R

> (a[3]).x    // OK: A Real[2] but not a component reference

Reviewer: Explain in terms of similar component reference?
```suggestion
(a[3]).x    // OK: Like a[3].x, but not a component reference
```

> (a[3]).x[1] // Error.

Reviewer: Looks like we're missing the final case:
```suggestion
(a[3]).x[1] // Error.
((a[3]).x)[1] // OK: same as a[3].x[1]
```

\subsection{Member Access Operator}\label{member-access-operator}

It is possible to access members of a class instance using dot notation, i.e., the \lstinline!.! operator.

> It is also possible to select a record member of a general expression by enclosing it in parenthesis.

Reviewer: I believe this should be the plural:
```suggestion
It is also possible to select a record member of a general expression by enclosing it in parentheses.
```

Note that while the selection is applied to an \lstinline[language=grammar]!output-expression-list! in the grammar, it is only semantically valid when the \lstinline[language=grammar]!output-expression-list! represents an expression.

In case the first operand is an array it is seen as a slicing operation, see \cref{slice-operation}.

\begin{example}

> \lstinline!R1.R! for accessing the resistance component \lstinline!R! of resistor \lstinline!R1!.

Reviewer: This needs to be rephrased in the context of a new and bigger example.
```suggestion
\lstinline!R1.R! is a component reference accessing the resistance component \lstinline!R! of resistor \lstinline!R1!.

```

> Another use of dot notation: local classes which are members of a class can of course also be accessed using dot notation on the name of the class, not on instances of the class.

Reviewer: Similar to above:
```suggestion
The qualified class name \lstinline!A.B! is a reference to the local class \lstinline!B! which is a member of the class \lstinline!A!.
Note that the left operand in this case is a class, not an instance of the class.
```

\subsection{Member Access Operator}\label{member-access-operator}

It is possible to access members of a class instance using dot notation, i.e., the \lstinline!.! operator.
It is also possible to select a record member of a general expression by enclosing it in parentheses.
Note that while the selection is applied to an \lstinline[language=grammar]!output-expression-list! in the grammar, it is only semantically valid when the \lstinline[language=grammar]!output-expression-list! represents an expression.

In case the first operand is an array it is seen as a slicing operation, see \cref{slice-operation}.

\begin{example}

> The component reference \lstinline!R1.R! accesses the resistance component \lstinline!R! of resistor \lstinline!R1!.

Reviewer: Lower camel case of instance names would further help separating this from the form in next paragraph, but I don't know what to do with the fact that `R` is such a well established name for the resistance parameter…

Author: We will have to improve that later. I guess we should instance rename the component `R1`

\subsection{Member Access Operator}\label{member-access-operator}

It is possible to access members of a class instance using dot notation, i.e., the \lstinline!.! operator.
It is also possible to select a record member of a general expression by enclosing it in parenthesis.

> Note that while the selection is applied to a output-expression-list in the grammar, it is only semantically valid when the output-expression-list represents an expression.

Reviewer: Typo and formatting:
```suggestion
Note that while the selection is applied to an \lstinline[language=grammar]!output-expression-list! in the grammar, it is only semantically valid when the \lstinline[language=grammar]!output-expression-list! represents an expression.
```

Same typo exists for indexing.

In case the first operand is an array it is seen as a slicing operation, see \cref{slice-operation}.

\begin{example}
\lstinline!R1.R! for accessing the resistance component \lstinline!R! of resistor \lstinline!R1!.
Another use of dot notation: local classes which are members of a class can of course also be accessed using dot notation on the name of the class, not on instances of the class.

\lstinline!(Complex(2, 3)).re! constructs the record \lstinline!Complex(2, 3)! and then selects the \lstinline!re! component in it.

> Using \lstinline!Complex(2, 3).re! is not syntactically legal.

Reviewer: Simpler:
```suggestion
\lstinline!Complex(2, 3).re! is not valid syntax.
```
